File: apps/api-python/api/validation.py
"""Ticket-readiness validator — the QA-input quality gate.

The /pyapi/generate endpoint and the /pyapi/validate-ticket endpoint both
delegate here. The flow is two-phase:

  1. Deterministic rules (cheap, fast, no LLM) — empty/short summary, empty
     or placeholder description, too few acceptance criteria, bug tickets
     missing repro/expected. Any rule failure produces a TicketValidationIssue
     and the rubric is SKIPPED (no point asking an LLM to score a ticket that
     already failed the obvious checks).
  2. LLM rubric (semantic) — only runs when rules pass and `use_llm_rubric`
     is True. The LLM acts as a senior QA reviewer and decides if the ticket
     is actually testable. See prompt.build_rubric_*_prompt for the contract.

Design contract: rubric infra failures (network, parse, schema) MUST NOT
block. We log, then return rules-only `passed=True`. This way a flaky model
provider never stops a structurally valid ticket from being generated.
"""

from __future__ import annotations

import json
import logging
import re
from typing import Iterable

from .llm import LLMCompletionRequest, LLMProvider, resolve_llm_provider
from .models import (
    IssueType,
    JiraTicket,
    Platform,
    TicketValidationIssue,
    TicketValidationResult,
)
from .prompt import build_rubric_system_prompt, build_rubric_user_prompt

logger = logging.getLogger("jiraqa.python.validation")


# ────────────────────────────────────────────────────────────────────────────
# Tunables
# ────────────────────────────────────────────────────────────────────────────

MIN_SUMMARY_LEN = 10
MIN_DESCRIPTION_LEN = 30
MIN_ACCEPTANCE_CRITERIA = 2
MIN_ACCEPTANCE_CRITERION_LEN = 5
RUBRIC_PASS_THRESHOLD = 70

# Whole-word placeholder tokens that signal an unfinished ticket. We match
# these as standalone words / phrases (case-insensitive) so a real word like
# "tba" inside a longer sentence wouldn't false-positive (`\b` boundaries).
_PLACEHOLDER_TOKENS = (
    "tbd",
    "todo",
    "tba",
    "n/a",
    "fill in later",
    "fill this in",
    "to be determined",
    "to be added",
)
_PLACEHOLDER_RE = re.compile(
    r"\b(" + "|".join(re.escape(t) for t in _PLACEHOLDER_TOKENS) + r")\b",
    flags=re.IGNORECASE,
)

# Words / phrases that indicate a bug ticket has steps-to-reproduce.
_REPRO_RE = re.compile(
    r"\b(steps? to reproduce|reproduce|reproduction|how to reproduce|repro steps?)\b",
    flags=re.IGNORECASE,
)
# Two halves of expected-vs-actual. We require BOTH to be present.
_EXPECTED_RE = re.compile(r"\b(expected|should|must)\b", flags=re.IGNORECASE)
_ACTUAL_RE = re.compile(
    r"\b(actual|got|but instead|currently|observed|happens)\b",
    flags=re.IGNORECASE,
)

# Summary "junk" patterns — single words / nonsense that are technically
# >= MIN_SUMMARY_LEN if you pad them, but obviously not a real ticket title.
_SUMMARY_JUNK_RE = re.compile(r"^(?:tbd|todo|test|asdf|xxx+|placeholder)$", re.IGNORECASE)


# ────────────────────────────────────────────────────────────────────────────
# Deterministic rules
#
# Each rule is a small pure function returning a list of issues. We collect
# them all in one pass so the QA team sees every problem at once instead of
# fixing-then-resubmitting-then-finding-the-next-one.
# ────────────────────────────────────────────────────────────────────────────


def _rule_summary(ticket: JiraTicket) -> list[TicketValidationIssue]:
    issues: list[TicketValidationIssue] = []
    s = (ticket.summary or "").strip()
    if not s:
        issues.append(
            TicketValidationIssue(
                field="summary",
                code="summary_missing",
                message="Summary is required.",
                hint="Add a one-line description of the change, e.g. 'Login screen rejects empty password with inline error'.",
            )
        )
        return issues
    if len(s) < MIN_SUMMARY_LEN:
        issues.append(
            TicketValidationIssue(
                field="summary",
                code="summary_too_short",
                message=f"Summary must be at least {MIN_SUMMARY_LEN} characters (got {len(s)}).",
                hint="Describe what changes and where, not just 'fix' or 'bug'.",
            )
        )
    if _SUMMARY_JUNK_RE.match(s):
        issues.append(
            TicketValidationIssue(
                field="summary",
                code="summary_placeholder",
                message=f"Summary '{s}' looks like a placeholder.",
                hint="Replace with a real one-line description of the change.",
            )
        )
    return issues


def _rule_description(ticket: JiraTicket) -> list[TicketValidationIssue]:
    issues: list[TicketValidationIssue] = []
    d = (ticket.description or "").strip()
    if not d:
        issues.append(
            TicketValidationIssue(
                field="description",
                code="description_missing",
                message="Description is required.",
                hint="Explain the desired behaviour, screen/component, and any constraints.",
            )
        )
        return issues
    if len(d) < MIN_DESCRIPTION_LEN:
        issues.append(
            TicketValidationIssue(
                field="description",
                code="description_too_short",
                message=f"Description must be at least {MIN_DESCRIPTION_LEN} characters (got {len(d)}).",
                hint="Add context: what changes, where, why, and any edge cases to consider.",
            )
        )
    placeholder = _PLACEHOLDER_RE.search(d)
    if placeholder:
        issues.append(
            TicketValidationIssue(
                field="description",
                code="description_placeholder",
                message=f"Description contains placeholder text ('{placeholder.group(1)}'). Replace it with the actual behaviour.",
                hint="What should happen? Where? Under what conditions?",
            )
        )
    return issues


def _rule_acceptance_criteria(ticket: JiraTicket) -> list[TicketValidationIssue]:
    issues: list[TicketValidationIssue] = []
    acs = [a.strip() for a in (ticket.acceptance_criteria or []) if a and a.strip()]
    if len(acs) < MIN_ACCEPTANCE_CRITERIA:
        issues.append(
            TicketValidationIssue(
                field="acceptance_criteria",
                code="ac_too_few",
                message=f"Need at least {MIN_ACCEPTANCE_CRITERIA} acceptance criteria (got {len(acs)}).",
                hint="Add at least a happy-path case and a negative / edge case.",
            )
        )
    # Per-criterion minimum length — catches ["x", "y"] passing the count.
    short = [a for a in acs if len(a) < MIN_ACCEPTANCE_CRITERION_LEN]
    if short:
        issues.append(
            TicketValidationIssue(
                field="acceptance_criteria",
                code="ac_too_short",
                message=f"{len(short)} acceptance criteria are too short to be testable.",
                hint="Use Given/When/Then phrasing with concrete inputs and outputs.",
            )
        )
    return issues


def _rule_bug_specifics(ticket: JiraTicket) -> list[TicketValidationIssue]:
    """Bugs need both 'steps to reproduce' and 'expected vs actual'.

    Skipped for non-bug issue types. We look in the description only — the
    summary is too short to carry both halves of expected-vs-actual.
    """
    if ticket.issue_type != IssueType.bug:
        return []
    issues: list[TicketValidationIssue] = []
    d = ticket.description or ""
    if not _REPRO_RE.search(d):
        issues.append(
            TicketValidationIssue(
                field="description",
                code="bug_missing_repro",
                message="Bug tickets must include steps to reproduce.",
                hint="Add a 'Steps to reproduce:' section with numbered steps.",
            )
        )
    if not (_EXPECTED_RE.search(d) and _ACTUAL_RE.search(d)):
        issues.append(
            TicketValidationIssue(
                field="description",
                code="bug_missing_expected_vs_actual",
                message="Bug tickets must describe expected vs actual behaviour.",
                hint="Add 'Expected: ...' and 'Actual: ...' lines.",
            )
        )
    return issues


_ALL_RULES = (
    _rule_summary,
    _rule_description,
    _rule_acceptance_criteria,
    _rule_bug_specifics,
)


def run_deterministic_rules(ticket: JiraTicket) -> list[TicketValidationIssue]:
    """Run every rule and aggregate issues. Order is stable so tests can rely on it."""
    issues: list[TicketValidationIssue] = []
    for rule in _ALL_RULES:
        issues.extend(rule(ticket))
    return issues


# ────────────────────────────────────────────────────────────────────────────
# LLM rubric pass
# ────────────────────────────────────────────────────────────────────────────


async def _run_rubric(
    ticket: JiraTicket,
    platform: Platform,
    llm: LLMProvider,
) -> tuple[int | None, str | None, list[TicketValidationIssue]]:
    """Call the LLM rubric and parse its JSON response.

    Returns (score, summary, issues). On any infra/parse error returns
    (None, None, []) — callers treat that as "rubric skipped" and don't block.
    """
    system = build_rubric_system_prompt()
    user = build_rubric_user_prompt(ticket, platform)
    try:
        resp = await llm.complete(
            LLMCompletionRequest(
                system_prompt=system,
                user_prompt=user,
                temperature=0.1,  # rubric scoring should be stable across runs
                json_mode=True,
                max_tokens=800,
            )
        )
    except Exception as e:  # noqa: BLE001
        logger.warning("Rubric LLM call failed (continuing without rubric): %s", e)
        return None, None, []

    try:
        parsed = json.loads(resp.text)
    except json.JSONDecodeError as e:
        logger.warning("Rubric LLM returned invalid JSON (continuing without rubric): %s", e)
        return None, None, []

    score_raw = parsed.get("score")
    try:
        score = int(score_raw) if score_raw is not None else None
    except (TypeError, ValueError):
        score = None
    summary = parsed.get("summary") if isinstance(parsed.get("summary"), str) else None

    raw_issues: Iterable[dict] = parsed.get("issues") or []
    parsed_issues: list[TicketValidationIssue] = []
    for it in raw_issues:
        if not isinstance(it, dict):
            continue
        try:
            parsed_issues.append(
                TicketValidationIssue(
                    field=str(it.get("field", "ticket")),
                    code=str(it.get("code", "rubric_issue")),
                    severity=str(it.get("severity", "error")),
                    message=str(it.get("message", "")),
                    hint=it.get("hint") if isinstance(it.get("hint"), str) else None,
                )
            )
        except Exception as e:  # noqa: BLE001
            logger.debug("Skipping malformed rubric issue: %s (%s)", it, e)

    # Trust the threshold, not the LLM's `passed` flag — keeps behaviour
    # consistent if the model returns score=65 but passed=true.
    if score is not None and score < RUBRIC_PASS_THRESHOLD and not parsed_issues:
        # Score is low but no issues listed — synthesize one so the user
        # at least sees the verdict instead of a silent rejection.
        parsed_issues.append(
            TicketValidationIssue(
                field="ticket",
                code="rubric_low_score",
                message=summary or f"Ticket scored {score}/100 on the QA-readiness rubric.",
                hint="Tighten expected behaviour, add measurable thresholds, and include edge cases.",
            )
        )

    return score, summary, parsed_issues


# ────────────────────────────────────────────────────────────────────────────
# Public entry point
# ────────────────────────────────────────────────────────────────────────────


async def validate_ticket(
    ticket: JiraTicket,
    platform: Platform,
    *,
    use_llm_rubric: bool = True,
    provider: str | None = None,
) -> TicketValidationResult:
    """Run the full validation pipeline. See module docstring for the contract."""
    # Phase 1 — deterministic rules.
    rule_issues = run_deterministic_rules(ticket)
    if rule_issues:
        # Don't waste an LLM call on a ticket that already failed obvious checks.
        return TicketValidationResult(passed=False, issues=rule_issues)

    # Phase 2 — optional LLM rubric.
    if not use_llm_rubric:
        return TicketValidationResult(passed=True, issues=[])

    try:
        llm = resolve_llm_provider(provider)
    except Exception as e:  # noqa: BLE001
        # No provider configured. Treat as rubric-skipped, NOT as failure.
        logger.info("Rubric skipped — no LLM provider available: %s", e)
        return TicketValidationResult(passed=True, issues=[])

    score, summary, rubric_issues = await _run_rubric(ticket, platform, llm)

    passed = True
    if score is not None and score < RUBRIC_PASS_THRESHOLD:
        passed = False
    if rubric_issues:
        # Belt-and-braces: even if score is None or high, surface any issues
        # the rubric explicitly raised.
        passed = False

    return TicketValidationResult(
        passed=passed,
        issues=rubric_issues,
        rubric_score=score,
        rubric_summary=summary,
    )

File: packages/core/src/validation.ts
/**
 * @module @jiraqa/core/validation
 *
 * Ticket-readiness validator — TypeScript mirror of apps/api-python/api/validation.py.
 *
 * Keep this file in lockstep with the Python version. Both backends MUST
 * produce the same `issues` list for the same ticket; the rubric pass is
 * the only place behaviour legitimately diverges (different LLMs).
 *
 * The deterministic rules live here so:
 *   - The TS /api/generate route can gate without crossing to the Python sidecar.
 *   - The UI can do an instant client-side preview (no round-trip) if it wants.
 *   - The standalone /api/validate-ticket route can serve TS-backend users.
 */

import { z } from "zod";
import { JiraTicketSchema, PlatformSchema } from "./types";
import type { JiraTicket, Platform } from "./types";

// ────────────────────────────────────────────────────────────────────────────
// Tunables — keep in sync with apps/api-python/api/validation.py
// ────────────────────────────────────────────────────────────────────────────

export const MIN_SUMMARY_LEN = 10;
export const MIN_DESCRIPTION_LEN = 30;
export const MIN_ACCEPTANCE_CRITERIA = 2;
export const MIN_ACCEPTANCE_CRITERION_LEN = 5;
export const RUBRIC_PASS_THRESHOLD = 70;

const PLACEHOLDER_TOKENS = [
  "tbd",
  "todo",
  "tba",
  "n/a",
  "fill in later",
  "fill this in",
  "to be determined",
  "to be added",
] as const;
// \b word boundaries — won't match inside longer words.
const PLACEHOLDER_RE = new RegExp(
  `\\b(${PLACEHOLDER_TOKENS.map((t) => t.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")).join("|")})\\b`,
  "i",
);

const REPRO_RE = /\b(steps? to reproduce|reproduce|reproduction|how to reproduce|repro steps?)\b/i;
const EXPECTED_RE = /\b(expected|should|must)\b/i;
const ACTUAL_RE = /\b(actual|got|but instead|currently|observed|happens)\b/i;
const SUMMARY_JUNK_RE = /^(?:tbd|todo|test|asdf|xxx+|placeholder)$/i;

// ────────────────────────────────────────────────────────────────────────────
// Zod schemas — wire shape for /api/validate-ticket
// ────────────────────────────────────────────────────────────────────────────

export const TicketValidationIssueSchema = z.object({
  field: z.string(), // "summary" | "description" | "acceptance_criteria" | "ticket"
  code: z.string(),
  severity: z.string().default("error"),
  message: z.string(),
  hint: z.string().optional().nullable(),
});
export type TicketValidationIssue = z.infer<typeof TicketValidationIssueSchema>;

export const TicketValidationResultSchema = z.object({
  passed: z.boolean(),
  issues: z.array(TicketValidationIssueSchema).default([]),
  rubric_score: z.number().int().nullable().optional(),
  rubric_summary: z.string().nullable().optional(),
});
export type TicketValidationResult = z.infer<typeof TicketValidationResultSchema>;

export const ValidateTicketRequestSchema = z.object({
  ticket: JiraTicketSchema,
  platform: PlatformSchema,
  use_llm_rubric: z.boolean().optional().default(true),
  provider: z.enum(["openai", "gemini", "anthropic"]).optional(),
});
export type ValidateTicketRequest = z.infer<typeof ValidateTicketRequestSchema>;

// ────────────────────────────────────────────────────────────────────────────
// Rules
// ────────────────────────────────────────────────────────────────────────────

function ruleSummary(ticket: JiraTicket): TicketValidationIssue[] {
  const issues: TicketValidationIssue[] = [];
  const s = (ticket.summary ?? "").trim();
  if (!s) {
    issues.push({
      field: "summary",
      code: "summary_missing",
      severity: "error",
      message: "Summary is required.",
      hint: "Add a one-line description of the change, e.g. 'Login screen rejects empty password with inline error'.",
    });
    return issues;
  }
  if (s.length < MIN_SUMMARY_LEN) {
    issues.push({
      field: "summary",
      code: "summary_too_short",
      severity: "error",
      message: `Summary must be at least ${MIN_SUMMARY_LEN} characters (got ${s.length}).`,
      hint: "Describe what changes and where, not just 'fix' or 'bug'.",
    });
  }
  if (SUMMARY_JUNK_RE.test(s)) {
    issues.push({
      field: "summary",
      code: "summary_placeholder",
      severity: "error",
      message: `Summary '${s}' looks like a placeholder.`,
      hint: "Replace with a real one-line description of the change.",
    });
  }
  return issues;
}

function ruleDescription(ticket: JiraTicket): TicketValidationIssue[] {
  const issues: TicketValidationIssue[] = [];
  const d = (ticket.description ?? "").trim();
  if (!d) {
    issues.push({
      field: "description",
      code: "description_missing",
      severity: "error",
      message: "Description is required.",
      hint: "Explain the desired behaviour, screen/component, and any constraints.",
    });
    return issues;
  }
  if (d.length < MIN_DESCRIPTION_LEN) {
    issues.push({
      field: "description",
      code: "description_too_short",
      severity: "error",
      message: `Description must be at least ${MIN_DESCRIPTION_LEN} characters (got ${d.length}).`,
      hint: "Add context: what changes, where, why, and any edge cases to consider.",
    });
  }
  const m = d.match(PLACEHOLDER_RE);
  if (m) {
    issues.push({
      field: "description",
      code: "description_placeholder",
      severity: "error",
      message: `Description contains placeholder text ('${m[1]}'). Replace it with the actual behaviour.`,
      hint: "What should happen? Where? Under what conditions?",
    });
  }
  return issues;
}

function ruleAcceptanceCriteria(ticket: JiraTicket): TicketValidationIssue[] {
  const issues: TicketValidationIssue[] = [];
  const acs = (ticket.acceptance_criteria ?? [])
    .map((a) => (a ?? "").trim())
    .filter((a) => a.length > 0);
  if (acs.length < MIN_ACCEPTANCE_CRITERIA) {
    issues.push({
      field: "acceptance_criteria",
      code: "ac_too_few",
      severity: "error",
      message: `Need at least ${MIN_ACCEPTANCE_CRITERIA} acceptance criteria (got ${acs.length}).`,
      hint: "Add at least a happy-path case and a negative / edge case.",
    });
  }
  const short = acs.filter((a) => a.length < MIN_ACCEPTANCE_CRITERION_LEN);
  if (short.length > 0) {
    issues.push({
      field: "acceptance_criteria",
      code: "ac_too_short",
      severity: "error",
      message: `${short.length} acceptance criteria are too short to be testable.`,
      hint: "Use Given/When/Then phrasing with concrete inputs and outputs.",
    });
  }
  return issues;
}

function ruleBugSpecifics(ticket: JiraTicket): TicketValidationIssue[] {
  if (ticket.issue_type !== "bug") return [];
  const issues: TicketValidationIssue[] = [];
  const d = ticket.description ?? "";
  if (!REPRO_RE.test(d)) {
    issues.push({
      field: "description",
      code: "bug_missing_repro",
      severity: "error",
      message: "Bug tickets must include steps to reproduce.",
      hint: "Add a 'Steps to reproduce:' section with numbered steps.",
    });
  }
  if (!(EXPECTED_RE.test(d) && ACTUAL_RE.test(d))) {
    issues.push({
      field: "description",
      code: "bug_missing_expected_vs_actual",
      severity: "error",
      message: "Bug tickets must describe expected vs actual behaviour.",
      hint: "Add 'Expected: ...' and 'Actual: ...' lines.",
    });
  }
  return issues;
}

/**
 * Run every deterministic rule and aggregate issues. Used by both /api/generate
 * (as a gate) and /api/validate-ticket (as the first phase).
 */
export function runDeterministicRules(ticket: JiraTicket): TicketValidationIssue[] {
  return [
    ...ruleSummary(ticket),
    ...ruleDescription(ticket),
    ...ruleAcceptanceCriteria(ticket),
    ...ruleBugSpecifics(ticket),
  ];
}

/**
 * Convenience wrapper that returns a full TicketValidationResult. The LLM
 * rubric pass is NOT performed here — the rubric lives in the route handler
 * so it can use the right provider abstraction. Phase-1 only.
 */
export function validateTicketRulesOnly(ticket: JiraTicket): TicketValidationResult {
  const issues = runDeterministicRules(ticket);
  return { passed: issues.length === 0, issues };
}

// ────────────────────────────────────────────────────────────────────────────
// Rubric prompts — TS mirror of prompt.build_rubric_*_prompt in Python.
// ────────────────────────────────────────────────────────────────────────────

export function buildRubricSystemPrompt(): string {
  return `You are a senior QA reviewer. Your job is to decide whether a Jira ticket
is TESTABLE — i.e. whether a QA engineer could write a step-by-step test case
from it without guessing.

Score the ticket on a 0-100 scale across these dimensions:
  1. Clarity of expected behaviour — Are post-conditions measurable, or vague
     ("snappier", "more reliable", "better UX")?
  2. Acceptance criteria quality — Are they Given/When/Then-style with concrete
     inputs and outputs, or do they restate the summary?
  3. Preconditions / environment — Is the starting state specified well enough
     to reproduce the scenario?
  4. Edge / negative coverage — Does the ticket hint at boundaries, errors,
     empty states, or only the happy path?

Threshold: score >= ${RUBRIC_PASS_THRESHOLD} ⇒ passed=true. Score < ${RUBRIC_PASS_THRESHOLD} ⇒ passed=false and you MUST
populate \`issues\` explaining what to fix.

OUTPUT
Return ONLY a JSON object matching exactly this shape, no prose, no fences:
{
  "score": <integer 0-100>,
  "passed": <boolean>,
  "summary": "<one-sentence verdict>",
  "issues": [
    {
      "field": "summary" | "description" | "acceptance_criteria" | "ticket",
      "code": "<snake_case identifier, e.g. ambiguous_expected_behavior>",
      "severity": "error",
      "message": "<what's wrong in plain English>",
      "hint": "<concrete suggestion for how to fix it>"
    }
  ]
}
If passed=true, \`issues\` MUST be an empty list.`;
}

export function buildRubricUserPrompt(ticket: JiraTicket, platform: Platform): string {
  const lines: string[] = [];
  lines.push("# JIRA TICKET TO REVIEW");
  lines.push(`Ticket ID: ${ticket.ticket_id}`);
  lines.push(`Summary: ${ticket.summary}`);
  lines.push(`Issue Type: ${ticket.issue_type}`);
  lines.push(`Priority: ${ticket.priority}`);
  if (ticket.component) lines.push(`Component: ${ticket.component}`);
  if (ticket.environment) lines.push(`Environment: ${ticket.environment}`);
  if (ticket.labels && ticket.labels.length > 0) {
    lines.push(`Labels: ${ticket.labels.join(", ")}`);
  }
  lines.push("");
  lines.push("Description:");
  lines.push(ticket.description || "(empty)");
  lines.push("");
  lines.push("Acceptance Criteria:");
  if (ticket.acceptance_criteria && ticket.acceptance_criteria.length > 0) {
    ticket.acceptance_criteria.forEach((ac, i) => lines.push(`  ${i + 1}. ${ac}`));
  } else {
    lines.push("  (none)");
  }
  lines.push("");
  lines.push(`Target platform: ${platform}`);
  lines.push("");
  lines.push("Return ONLY the JSON object as specified in the system prompt.");
  return lines.join("\n");
}
